### DISCUSSION TITLE
How could I tell different rules with the same prefix?

### DISCUSSION THREAD

**Discussion by shuizaiku7** [taocpp/PEGTL#332] [Q&A] [answered]
```
struct integer : pegtl::plus<pegtl::digit> {};
struct lexFloat : pegtl::seq<pegtl::star<pegtl::digit>, pegtl::one<'.'>, pegtl::plus<pegtl::digit>> {};
struct grammar : pegtl::must<pegtl::sor<integer, lexFloat>> {};
```

if the input is 1.2, it will match integer rule, namely just match 1

I've got 2 ways to address this.
01.Carefully select the order in sor
02.Match interger and lexFloat along with its following space

Is there any simpler way to solve this problem?

**Comment by ColinH** [comment, marked as answer]
Given the deterministic semantics of PEGs, which is a fancy way of saying that `sor<>` goes from left to right and stops at the first sub-rule that matched (even if a later rule would be a "better" match), it is normal and expected to carefully order the sub-rules to avoid this prefix problem you encountered.

Now in general it's a good idea to factor out common prefixes, so instead of `sor< seq< A, B >, seq< A, C >` you'd write `seq< A, sor< B, C > >`, however with your rule it is not quite that simple and while you could rewrite it as

```if_then_else< one< '.' >, plus< digit >, seq< plus< digit >, opt< one< '.' >, plus< digit > > > >```

you might prefer sticking to your version, though with a couple of "renamed" intermediate rules you can make it clearer which role each set of digits is playing and simplify attaching your actions:

```
struct dot : one< '.' >;
struct int_digits : plus< digit >{};
struct frac_digits : plus< digit >{};

struct number : if_then_else< dot, frac_digits, seq< int_digits, opt< dot, frac_digits > > > {};
```

https://github.com/taocpp/PEGTL/blob/main/doc/Rule-Reference.md#if_then_else-r-s-t-

**Reply by shuizaiku7** [reply]
Yeah, I think the work of tell different tokens that token A may be the prefix of token B in hand-writen parser has been done by lexer.
Cause pegtl will handle lexer/parser together, so we need to handle it on our own.
01.Just distinguish at every sor, put the longer rule before others.
02.Or provide more intermediate rules so that it's not possible for token A to be prefix of token B.

Thanks.
